### ISSUE TITLE
SCIDE attempts to open .zip archives

### ISSUE THREAD

**Issue by jamshark70** [supercollider/supercollider#4634]
## Environment

* SuperCollider version: develop
* Operating system: Ubuntu
* Other details (Qt version, audio driver, etc.):

## Steps to reproduce

Attempt to open a .zip file in the IDE.

## Expected vs. actual behavior

Expected behavior: IDE should ask "are you sure?"

Actual behavior: If it's a small .zip, the IDE displays garbage. If it's a large .zip, IDE hangs and must be force-killed.

**Comment by gusano** [comment]
I would argue that opening a .zip should not even be possible

**Comment by brianlheim** [comment]
this is user error, IMO. the only thing i see objectionable is that the IDE should perhaps warn if the file to be opened is very large and will stall the editor. otherwise, why have an opinion on filetypes? most other text editors will happily open any file you point them at, including those in a binary format like zip.

**Comment by jamshark70** [comment]
FWIW the editor *does* pop up a warning box if you try to open a file with an unfamiliar extension, even if it's really a text file -- e.g., I save a TX Modular patch as something.txmod (which internally is SC code actually), and open it in the IDE, and the IDE says "this might be a binary file, are you sure you want to open it?"

But if it's a zip file, and almost certainly binary, there is no warning and the IDE just tries to open it (and presumably tokenize it, which I bet is what looks like a hang).

So why does this file type slip through the cracks?

**Comment by dyfer** [comment]
On macOS (and current dev build) I don't get any warnings when trying to open other files - I tried .zip, .o, .scx and a binary executable (supernova)... IDE opened all of them (displaying garbage of course).

**Comment by muellmusik** [comment]
It would be easy to check extension in DocumentManager::open (and I assume that's the correct way to do this). I'm happy to make a PR for this, but can I just confirm all valid file types: scd, sc, schelp, txt, and rtf? Any additional plain text formats in other platforms?

Currently the code seems to assume that all files which aren't sc formats or rtf are plaintext, which doesn't seem sensible.

Would probably also be worth it to add a filter for types to the QFileDialog call for files opened from the menu action.

**Comment by telephon** [comment]
I've sometime edited `md` and `html` and `ly`(lilypond) files in the IDE. Most text editors will open an arbitrary file, IIRC.

**Comment by muellmusik** [comment]
> Most text editors will open an arbitrary file, IIRC.

True and that is occasionally useful. Still that seems more sensible in a general purpose editor rather than a lang-specific one like the IDE. What's the feeling? Feature or bug? If the former we can close this and leave it to users.

**Comment by telephon** [comment]
My suspicion is that people use the IDE in all kind of ways, and we may break their workflows.

**Comment by muellmusik** [comment]
Okay, well it does seem something that users can control (ie by not opening zip files) so we could close if @jamshark70 and @dyfer are okay with that. Alternatively we could try to stop the hang at least, assuming that’s not just a slow load.

**Comment by telephon** [comment]
Yes, a warning if you atttempt to open a really large "text file" might be good.

**Comment by jamshark70** [comment]
> Okay, well it does seem something that users can control (ie by not opening zip files) so we could close...

I discovered this behavior when my touchpad misread a double-click -- i.e. an accident. If accidents were 100% preventable by human agency, then we wouldn't need car insurance -- so I'm not quite convinced that we should summarily close.

I don't object actually to *opening* the file. I do object to whatever processing was going on that caused the IDE to spin up to 100% CPU, requiring a force-kill to regain control. Passing garbage into the code tokenizer might well cause the behavior, though I didn't investigate so it could be something else.

If we can reasonably anticipate that opening a certain file will lock up the editor for an hour or more, I think it would be polite at least to pop up an "are you sure"?

*Also*, if you open a text file with an unrecognized extension, SC *does* warn you that "it might be binary" but if it's a zip file (known to be binary) then it blindly opens the file and destroys your session. Let's at least not leave it at that!

**Comment by dyfer** [comment]
IMO we should make sure the warning about opening a binary file is always present.

Upon further check:
- the IDE asks "are you sure you want to open this" only if I drag and drop a .zip file, 
  - when I use the File -> Open dialog, it does _not_ ask and opens the file
- IDE did not hang with the files I tried (including a 40MB zip)
- I don't think strictly preventing opening files with weird extensions is the best way and I agree that the user might want to edit arbitrary files (including no extension)

(I'm on macOS, I wonder if any of this is platform-specific)

The IDE hang should still be investigated I think. @jamshark70 does it happen for all zip files on your system? This should be probably investigated further.

> It would be easy to check extension in DocumentManager::open (and I assume that's the correct way to do this)

I seem to remember back in the day (?) some linux distros insisted on recognizing file type by MIME type and not necessarily by the extension, but my memory is fuzzy about that.
